### PR TITLE
Implement login with Trezor Model T - Closes #1846

### LINKED ISSUE #1846: Implement login with Trezor Model T
This is a subtask of #1319. 
In this task, it should not be available to end users yet (e.g. set some local storage flag). There will be a final task to enable it.

The goal is to implement login with Trezor Model T.

### PR DESCRIPTION
**Pull request by michaeltomasik**
Implement login trezor model

### What issue have I solved?

#1846

### How have I implemented/fixed it?

1. First, install Trezor bridge https://wallet.trezor.io/#/bridge
2. localStorage.setItem('trezor', true) to unlock trezor login
3. Try to log in with trezor

### Review checklist
- The PR follows our [Test guide](/LiskHQ/lisk-hub/blob/development/docs/TEST_GUIDE.md)
- The PR follows our [CSS guide](/LiskHQ/lisk-hub/blob/development/docs/CSS_GUIDE.md)

### CODE REVIEW

File: app/src/trezor.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  }
+
+  win.send({ event: 'trezorPassphraseCallback', value: null });
+
+  ipcMain.once('trezorPassphraseCallbackResponse', (event, passphrase) => {
+    if (passphrase || passphrase === '') {
+      callback(null, passphrase);
+      // Save passphrase in case of Trezor One (it asks on every action. Pretty annoying)
+      const trezorDevice = getDeviceById(deviceId);
+      if (trezorDevice && trezorDevice.model === 'Trezor One') {
```
Comment (osvaldovega) [inline app/src/trezor.js]: I think is better define all information related to HW in a constants file, then in the future if we want to update something will be only in the constants file

File: app/src/trezor.js [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+  // Filter out devices not initialized
+  if (!device.isInitialized()) {
+    logDebug(`Device ${device.features.label} is not initialized.`);
+    win.send({ event: 'trezorParamMessage', value: 'TREZOR_IS_NOT_INITIALIZED' });
+    return;
+  }
+
+  // Filter out devices with unsupported firmware versions
+  if (device.features.model === '1' && !device.atLeast('1.7.1')) {
```
Comment (osvaldovega) [inline app/src/trezor.js]: the versions too

File: app/src/trezor.js [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    logDebug(`Device ${device.features.label} firmware version not supported.`);
+    win.send({ event: 'trezorParamMessage', value: 'TREZOR_ONE_OLD_FIRMWARE' });
+    return;
+  }
+  if (device.features.model === 'T' && !device.atLeast('2.0.7')) {
```
Comment (osvaldovega) [inline app/src/trezor.js]: and here

File: app/src/utils.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { ipcMain } from 'electron'; // eslint-disable-line import/no-extraneous-dependencies
+import Lisk from 'lisk-elements'; // eslint-disable-line import/no-extraneous-dependencies
+
+export const createCommand = (k, fn) => {
```
Comment (osvaldovega) [inline app/src/utils.js]: change k to something more meaningful

File: app/src/utils.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  ipcMain.on(`${k}.request`, (event, ...args) => {
+    fn(...args)
+      .then(r => ({ success: true, data: r }))
+      .catch(e => ({ success: false, errorKey: e }))
+      .then(r => event.sender.send(`${k}.result`, r));
```
Comment (osvaldovega) [inline app/src/utils.js]: the parameters of the then and catch too.... no just r or e

File: app/src/utils.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  });
+};
+
+export const isValidAddress = address => address.length > 2 && address.length < 22 && address[address.length - 1] === 'L';
```
Comment (osvaldovega) [inline app/src/utils.js]: for this you can use the regex function that we have to see if it is a valid address.
is in src/utils/regex.js

File: src/actions/transactions.js [hunk 7]
```diff
@@ -12,7 +12,8 @@ import { getTimeOffset } from '../utils/hacks';
 import Fees from '../constants/fees';
 import transactionTypes from '../constants/transactionTypes';
 import { toRawLsk } from '../utils/lsk';
-import { sendWithLedger } from '../utils/api/ledger';
+// import { sendWithLedger } from '../utils/api/ledger';
```
Comment (osvaldovega) [inline src/actions/transactions.js]: you can remove the comment in case we dont need it

File: src/components/formattedNumber/index.js [hunk 8]
```diff
@@ -1,14 +1,16 @@
+/* eslint-disable */
+
 import React from 'react';
-import numeral from 'numeral';
-import * as locales from 'numeral/locales'; // eslint-disable-line
+// import numeral from 'numeral';
+// import * as locales from 'numeral/locales'; // eslint-disable-line
```
Comment (osvaldovega) [inline src/components/formattedNumber/index.js]: here too, remove the comments

File: src/components/formattedNumber/index.js [hunk 9]
```diff
@@ -1,14 +1,16 @@
...
 import { translate } from 'react-i18next';
 import i18n from '../../i18n';
 
 const FormattedNumber = ({ val }) => { // eslint-disable-line
   // set numeral language
-  numeral.locale(i18n.language);
-  const formatedNumber = numeral(val).format('0,0.[0000000000000]');
-  return <span>{formatedNumber}</span>;
+  // numeral.locale(i18n.language);
+  // const formatedNumber = numeral(val).format('0,0.[0000000000000]');
```
Comment (osvaldovega) [inline src/components/formattedNumber/index.js]: and here

File: src/components/hwWallet/ledgerLoginHOC.js [hunk 10]
```diff
@@ -10,7 +10,7 @@ import LedgerLogin from './ledgerLogin';
 const mapStateToProps = state => ({
   liskAPIClient: state.peers && state.peers.liskAPIClient,
   settings: state.settings,
-  loginType: state.account.loginType || 1,
+  // loginType: state.account.loginType || 1,
```
Comment (osvaldovega) [inline src/components/hwWallet/ledgerLoginHOC.js]: here too

File: src/components/hwWallet/trezorLogin.js [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+          trezorPasshpraseRequested: true,
+          trezorPinRequested: false,
+        });
+      });
+    }
+  }
+
+  async componentWillMount() {
+    // this.keyboardHandler = this.keyboardHandler.bind(this);
+    // window.addEventListener('keydown', this.keyboardHandler, false);
```
Comment (osvaldovega) [inline src/components/hwWallet/trezorLogin.js]: here too

File: src/components/hwWallet/trezorLoginHOC.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+import { liskAPIClientSet } from '../../actions/peers';
+import { settingsUpdated } from '../../actions/settings';
+import { errorToastDisplayed } from '../../actions/toaster';
+import TrezorLogin from './trezorLogin';
+
+const mapStateToProps = state => ({
+  liskAPIClient: state.peers && state.peers.liskAPIClient,
+  settings: state.settings,
+  // loginType: state.account.loginType || 1,
```
Comment (osvaldovega) [inline src/components/hwWallet/trezorLoginHOC.js]: remove comment

File: src/utils/api/hwWallet.js [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+
+  if (platform === PLATFORM_TYPES.ELECTRON) {
+    return sendIpcCommand(command);
+  }
+
+  if (platform === PLATFORM_TYPES.BROWSER) { // Used only during dev.
+    let resCommand;
+    if (command.hwType === 1) {
+      resCommand = await executeLedgerCommandForWeb(command);
+    } else if (command.hwType === 2) {
```
Comment (osvaldovega) [inline src/utils/api/hwWallet.js]: instead 1 and 2 use constants

File: src/utils/api/hwWallet.js [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+    }
+    return resCommand;
+  }
+  throw new Error(HW_MSG.NO_TRANSPORT_AVAILABLE);
+};
+
+export const getLoginTypeFromDevice = (device) => {
+  if (device.model === 'Ledger Nano S') {
+    return loginTypes.ledgerNano;
+  } else if (device.model === 'Trezor One' || device.model === 'Trezor Model T') {
```
Comment (osvaldovega) [inline src/utils/api/hwWallet.js]: instead use the string for each model use constants

File: src/utils/api/hwWallet.js [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+  // TODO Detach this from main process
+  // if (isInitialized) {
+  //   console.log('address', address);
+  //   const txAccount = await getTransactions(activePeer, address);
+  //   console.log('txAccount 2', txAccount);
+  //   Object.assign(resAccount, { txCount: txAccount.meta.count });
+
+  //   const votesAccount = await getVotes(activePeer, address);
+  //   Object.assign(resAccount, { votesCount: votesAccount.data.votesUsed });
+  // }
```
Comment (osvaldovega) [inline src/utils/api/hwWallet.js]: Do we need this code because if we dont maybe you can remove it

File: src/utils/hwWallet.js [hunk 16]
```diff
@@ -0,0 +1,303 @@
+/* eslint-disable */
+import TransportU2F from '@ledgerhq/hw-transport-u2f';
+import TrezorConnect from 'trezor-connect';
+import i18next from 'i18next';
+import { LedgerAccount, SupportedCoin, DposLedger } from 'dpos-ledger-api';
+import { HW_CMD, HW_MSG, calculateSecondPassphraseIndex } from '../constants/hwConstants';
+import { loadingStarted, loadingFinished } from './loading';
+// import { accountLoggedOut } from '../actions/account';
```
Comment (osvaldovega) [inline src/utils/hwWallet.js]: remove comment

**Comment by Efefefef** [comment]
Please transfer to Pending QA after all changes are finalized

**Pull request merged**